File: services/maps.py
from __future__ import annotations

import asyncio
from datetime import datetime, timedelta

import httpx

GEOCODING_URL = "https://maps.googleapis.com/maps/api/geocode/json"
PLACES_URL = "https://maps.googleapis.com/maps/api/place/autocomplete/json"
DIRECTIONS_URL = "https://maps.googleapis.com/maps/api/directions/json"


async def validate_address(address: str, api_key: str) -> tuple[bool, str | None]:
    async with httpx.AsyncClient() as client:
        resp = await client.get(GEOCODING_URL, params={"address": address, "key": api_key})
    data = resp.json()
    results = data.get("results", [])
    if data.get("status") == "OK" and results:
        return True, results[0]["formatted_address"]
    return False, None


async def suggest_address(query: str, api_key: str) -> list[str]:
    async with httpx.AsyncClient() as client:
        resp = await client.get(
            PLACES_URL,
            params={"input": query, "key": api_key, "types": "geocode"},
        )
    data = resp.json()
    if data.get("status") == "OK" and data.get("predictions"):
        return [p["description"] for p in data["predictions"][:3]]
    return []


async def _fetch_hourly(
    client: httpx.AsyncClient,
    origin: str,
    destination: str,
    hour: int,
    api_key: str,
    mode: str = "driving",
) -> tuple[int, list[dict]]:
    now = datetime.now()
    departure = now.replace(hour=hour, minute=0, second=0, microsecond=0)
    if departure < now:
        departure += timedelta(days=1)

    params: dict = {
        "origin": origin,
        "destination": destination,
        "key": api_key,
        "mode": mode,
        "departure_time": int(departure.timestamp()),
        "alternatives": "true",
    }
    if mode == "driving":
        params["traffic_model"] = "best_guess"

    resp = await client.get(DIRECTIONS_URL, params=params)
    data = resp.json()
    return hour, data.get("routes", [])


async def fetch_all_hourly_traffic(
    origin: str, destination: str, api_key: str, mode: str = "driving"
) -> dict[int, list[dict]]:
    async with httpx.AsyncClient(timeout=30) as client:
        tasks = [
            _fetch_hourly(client, origin, destination, hour, api_key, mode)
            for hour in range(24)
        ]
        results = await asyncio.gather(*tasks)
    return {hour: routes for hour, routes in results}


async def fetch_current_routes(
    origin: str, destination: str, api_key: str, mode: str = "driving"
) -> list[dict]:
    params: dict = {
        "origin": origin,
        "destination": destination,
        "key": api_key,
        "mode": mode,
        "alternatives": "true",
    }
    if mode == "driving":
        params["departure_time"] = "now"
        params["traffic_model"] = "best_guess"

    async with httpx.AsyncClient(timeout=15) as client:
        resp = await client.get(DIRECTIONS_URL, params=params)
    return resp.json().get("routes", [])

File: frontend/src/lib/maps.ts
const GEOCODING_URL = "https://maps.googleapis.com/maps/api/geocode/json";
const PLACES_URL = "https://maps.googleapis.com/maps/api/place/autocomplete/json";
const DIRECTIONS_URL = "https://maps.googleapis.com/maps/api/directions/json";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type GoogleRoute = Record<string, any>;

export async function validateAddress(
  address: string,
  apiKey: string,
): Promise<[boolean, string | null]> {
  const url = new URL(GEOCODING_URL);
  url.searchParams.set("address", address);
  url.searchParams.set("key", apiKey);
  const resp = await fetch(url.toString());
  const data = await resp.json();
  const results: GoogleRoute[] = data.results ?? [];
  if (data.status === "OK" && results.length) {
    return [true, results[0].formatted_address as string];
  }
  return [false, null];
}

export async function suggestAddress(
  query: string,
  apiKey: string,
): Promise<{ description: string; place_id: string }[]> {
  const url = new URL(PLACES_URL);
  url.searchParams.set("input", query);
  url.searchParams.set("key", apiKey);
  url.searchParams.set("language", "en");
  const resp = await fetch(url.toString());
  const data = await resp.json();
  if (data.status === "OK" && (data.predictions as GoogleRoute[])?.length) {
    return (data.predictions as GoogleRoute[]).slice(0, 5).map((p) => ({
      description: p.description as string,
      place_id: p.place_id as string,
    }));
  }
  return [];
}

async function fetchHourly(
  origin: string,
  destination: string,
  hour: number,
  apiKey: string,
  mode: string,
  tzOffsetMinutes: number,
): Promise<[number, GoogleRoute[]]> {
  // Compute Unix timestamp for `hour:00` in the client's local timezone.
  // tzOffsetMinutes = getTimezoneOffset() — positive for west, negative for east (e.g. SGT = -480).
  const nowMs = Date.now();
  const localNowMs = nowMs - tzOffsetMinutes * 60_000;
  const midnightLocalMs = Math.floor(localNowMs / 86_400_000) * 86_400_000;
  const targetUtcMs = midnightLocalMs + hour * 3_600_000 + tzOffsetMinutes * 60_000;
  const departureMs = targetUtcMs < nowMs ? targetUtcMs + 86_400_000 : targetUtcMs;
  const departureTime = Math.floor(departureMs / 1000);

  const url = new URL(DIRECTIONS_URL);
  url.searchParams.set("origin", origin);
  url.searchParams.set("destination", destination);
  url.searchParams.set("key", apiKey);
  url.searchParams.set("mode", mode);
  url.searchParams.set("departure_time", String(departureTime));
  url.searchParams.set("alternatives", "true");
  if (mode === "driving") url.searchParams.set("traffic_model", "best_guess");

  const resp = await fetch(url.toString());
  const data = await resp.json();
  const routes: GoogleRoute[] = data.routes ?? [];

  if (mode === "transit" && routes.length) {
    const actual: number = routes[0].legs?.[0]?.departure_time?.value ?? 0;
    if (actual - departureTime > 1800) return [hour, []];
  }

  return [hour, routes];
}

export async function fetchAllHourlyTraffic(
  origin: string,
  destination: string,
  apiKey: string,
  mode: string = "driving",
  tzOffsetMinutes: number = 0,
): Promise<Record<number, GoogleRoute[]>> {
  const hours = Array.from({ length: 19 }, (_, i) => i + 5); // 5..23
  const results = await Promise.all(
    hours.map((hour) => fetchHourly(origin, destination, hour, apiKey, mode, tzOffsetMinutes)),
  );
  return Object.fromEntries(results);
}
